samples/oauth_react/src/Authentication.jsx

```diff
@@ -0,0 +1,144 @@
+// This module exports the Authentication element to manage the various
+// openid-client objects required to log in and out. It also exports a variety
+// of elements to make it easier to consume the authentication context without
+// having to care about AuthenticationContext.
+
+import React, { Component } from 'react';
+import { Log, UserManager } from 'oidc-client';
+import { Button } from 'semantic-ui-react';
+
+const AuthenticationContext = React.createContext({user: null});
+
+class Authentication extends Component {
+  constructor(props) {
+    super(props);
+    this.state = {authenticated: false, user: null};
+    this.userManager = new UserManager(this.props.OidcSettings);
+    this.userManager.events.addUserLoaded((user) => {
+      console.log('user loaded, changing Authentication state');
+      this.setState({authenticated: true, user: user});
+    });
+    this.userManager.events.addUserUnloaded(() => {
+      console.log('user unloaded, changing Authentication state');
+      this.setState({authenticated: false, user: null});
+    });
+  }
+
+  componentDidMount() {
+    Log.logger = window.console;
+    this.userManager.getUser().then((user) => {
+      this.setState({authenticated: true, user: user});
+      console.log(user)
+    });
+  }
+
+  render() {
+    const ctxValue = {user: this.state.user, userManager: this.userManager, oidcSettings: this.props.oidcSettings};
+    return (
+      <AuthenticationContext.Provider value={ctxValue}>
+        {this.props.children}
+      </AuthenticationContext.Provider>
+    );
+  }
+}
+
+class ImplicitCallback extends Component {
+  static contextType = AuthenticationContext;
+
+  componentDidMount() {
+    this.context.userManager.signinRedirectCallback()
+      .then((user) => this.onRedirectSuccess(user))
+      .catch((error) => this.onRedirectError(error));
+  }
+
+  onRedirectSuccess(user) {
+    console.log(user);
+    this.props.history.push(`/`)
+  }
+
+  onRedirectError(error) {
+    console.log(error);
+  }
+
+  render() {
+    return null;
+  }
+}
+
+
+class AuthenticationState extends Component {
+  static contextType = AuthenticationContext;
+
+  constructor(props) {
+    super(props);
+    this.state = {
+      authenticated: false,
```

Comment:
Where is `authenticated` used? In `componentDidMount` it is used but only to set itself to a new value. Is it used to help show authentication state to developers trying to understand the login process? Seems like `user` being null or not determines login state, not `authenticated`.

Reply:
Excellent question :smile: I think this is left behind from an earlier approach. When I started out I thought I could avoid a bunch of these components reaching up the tree for an `AutenticationContext.Provider` by caching this state within `SignedIn` and `SignedOut`. That approach worked out poorly because multiple components needed to use the same empty state detection, effectively coupling them together. When I first started with that approach I was relying on the `componentWillMount` lifecycle callback but then I learned that [it is deprecated and will be removed in react v17](https://reactjs.org/docs/react-component.html#legacy-lifecycle-methods). I'll clean this up soon.